driver-core/src/main/com/mongodb/connection/DefaultAuthenticator.java

```diff
@@ -16,33 +16,95 @@
+            }
+        }
+        authenticator.authenticate(connection, connectionDescription);
     }
 
     @Override
     void authenticateAsync(final InternalConnection connection, final ConnectionDescription connectionDescription,
                            final SingleResultCallback<Void> callback) {
-        createAuthenticator(connectionDescription).authenticateAsync(connection, connectionDescription, callback);
+        if (connectionDescription.getServerVersion().compareTo(FOUR_ZERO) < 0) {
+            getLegacyDefaultAuthenticator(connectionDescription).authenticateAsync(connection, connectionDescription, callback);
+        } else {
+            executeCommandAsync("admin", createIsMasterCommand(), connection, new SingleResultCallback<BsonDocument>() {
+                @Override
+                public void onResult(final BsonDocument result, final Throwable t) {
+                    if (t != null) {
+                        callback.onResult(null, wrapException(t));
+                    } else {
+                        getAuthenticatorFromIsMasterResult(result).authenticateAsync(connection, connectionDescription, callback);
+                    }
+                }
+            });
+        }
+    }
+
+    private Authenticator getAuthenticatorFromIsMasterResult(final BsonDocument isMasterResult) {
+        BsonArray saslSupportedMechs = isMasterResult.getArray("saslSupportedMechs", DEFAULT_SUPPORTED_MECHANISMS);
+        AuthenticationMechanism mechanism = saslSupportedMechs.contains(DEFAULT_MECHANISM_NAME) ? SCRAM_SHA_256 : SCRAM_SHA_1;
+        return new ScramShaAuthenticator(getCredential(), mechanism);
     }
 
-    Authenticator createAuthenticator(final ConnectionDescription connectionDescription) {
-        if (connectionDescription.getServerVersion().compareTo(new ServerVersion(2, 7)) >= 0) {
-            return new ScramSha1Authenticator(getCredential());
+    private Authenticator getLegacyDefaultAuthenticator(final ConnectionDescription connectionDescription) {
+        if (connectionDescription.getServerVersion().compareTo(THREE_ZERO) >= 0) {
+            return new ScramShaAuthenticator(getCredential(), SCRAM_SHA_1);
         } else {
             return new NativeAuthenticator(getCredential());
         }
     }
+
+    private BsonDocument createIsMasterCommand() {
+        BsonDocument isMasterCommandDocument = new BsonDocument("ismaster", new BsonInt32(1));
+        isMasterCommandDocument.append("saslSupportedMechs",
+                new BsonString(format("%s.%s", getCredential().getSource(), getCredential().getUserName())));
+        return isMasterCommandDocument;
+    }
+
+    private MongoSecurityException wrapException(final Throwable t) {
+        if (t instanceof MongoSecurityException) {
+            return (MongoSecurityException) t;
+        } else {
+            return new MongoSecurityException(getCredential(), format("Exception authenticating %s", getCredential()), t);
```

Comment:
Note: Still an open question on the auth spec PR on this related to whether a driver should look for a specific error code in the ismaster reply.

Reply:
Open question resolved: spec now says to switch on error code 11.